packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/CameraApiImpl.java

```diff
@@ -0,0 +1,411 @@
+  private final TextureRegistry textureRegistry;
+  private final EventChannel imageStreamChannel;
+  @VisibleForTesting @Nullable Camera camera;
+
+  CameraApiImpl(
+      Activity activity,
+      BinaryMessenger messenger,
+      CameraPermissions cameraPermissions,
+      PermissionsRegistry permissionsAdder,
+      TextureRegistry textureRegistry) {
+    this.activity = activity;
+    this.messenger = messenger;
+    this.cameraPermissions = cameraPermissions;
+    this.permissionsRegistry = permissionsAdder;
+    this.textureRegistry = textureRegistry;
+
+    imageStreamChannel =
+        new EventChannel(messenger, "plugins.flutter.io/camera_android/imageStream");
+    Messages.CameraApi.setUp(messenger, this);
+  }
+
+  void tearDownMessageHandler() {
+    Messages.CameraApi.setUp(messenger, null);
+  }
+
+  private Long instantiateCamera(String cameraName, Messages.PlatformMediaSettings settings)
+      throws CameraAccessException {
+    TextureRegistry.SurfaceTextureEntry flutterSurfaceTexture =
+        textureRegistry.createSurfaceTexture();
+    long cameraId = flutterSurfaceTexture.id();
+    DartMessenger dartMessenger =
+        new DartMessenger(
+            new Handler(Looper.getMainLooper()),
+            new Messages.CameraGlobalEventApi(messenger),
+            new Messages.CameraEventApi(messenger, String.valueOf(cameraId)));
+    CameraProperties cameraProperties =
+        new CameraPropertiesImpl(cameraName, CameraUtils.getCameraManager(activity));
+    Integer fps = (settings.getFps() == null) ? null : settings.getFps().intValue();
+    Integer videoBitrate =
+        (settings.getVideoBitrate() == null) ? null : settings.getVideoBitrate().intValue();
+    Integer audioBitrate =
+        (settings.getAudioBitrate() == null) ? null : settings.getAudioBitrate().intValue();
+    ResolutionPreset resolutionPreset =
+        CameraUtils.resolutionPresetFromPigeon(settings.getResolutionPreset());
+
+    camera =
+        new Camera(
+            activity,
+            flutterSurfaceTexture,
+            new CameraFeatureFactoryImpl(),
+            dartMessenger,
+            cameraProperties,
+            new Camera.VideoCaptureSettings(
+                resolutionPreset, settings.getEnableAudio(), fps, videoBitrate, audioBitrate));
+
+    return flutterSurfaceTexture.id();
+  }
+
+  // We move catching CameraAccessException out of onMethodCall because it causes a crash
+  // on plugin registration for sdks incompatible with Camera2 (< 21). We want this plugin to
+  // to be able to compile with <21 sdks for apps that want the camera and support earlier version.
+  @SuppressWarnings("ConstantConditions")
+  private <T> void handleException(Exception exception, Messages.Result<T> result) {
+    if (exception instanceof CameraAccessException) {
+      result.error(exception);
+      return;
+    }
+
+    // CameraAccessException can not be cast to a RuntimeException.
+    throw (RuntimeException) exception;
```

Comment:
Sorry, I realized looking at this again today that I was wrong about this; this is—non-obviously—a regression in this PR. With method channels, on Android, the engine call to `onMethodCall` [is wrapped in an exception handler](https://github.com/flutter/engine/blob/31aaaaad868743b38ac3b7165f0d58eca94e521b/shell/platform/android/io/flutter/plugin/common/MethodChannel.java#L285-L291) (specifically a `RuntimeException` handler, thus the current structure of this code). So the old code *would* correctly report these exceptions back to Dart, but this version won't because Pigeon doesn't `try/catch` `async` methods.

So we can close https://github.com/flutter/flutter/issues/157021, and this code needs to be updated here. To be fully behavior-preserving, `handleException` should look like this:


```
// The code below exactly preserves the format of the native exceptions generated by pre-Pigeon
// code. Since `camera` currently leaks the raw platform exceptions to clients, there may be client
// code that relies on specific string values here, so these should not be changed. See
// https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#platform-exception-handling
// for longer-term solutions to this.
if (exception instanceof CameraAccessException) {
  result.error(new FlutterError("CameraAccess", exception.getMessage(), null));
  return;
}
result.error(new FlutterError("error", exception.getMessage(), null, Log.getStackTraceString(exception)));
```

Apologies for missing this before; I'm still more used to the Obj-C method channel behaviors that the Java method channels.